File: easy/201-300/248/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#[derive(Debug, PartialEq, Eq, Clone)]
struct Color {
    red: u8,
    green: u8,
    blue: u8,
}

struct Grid {
    width: u32,
    height: u32,
    points: Vec<Color>,
}

impl Grid {
    fn new(width: u32, height: u32) -> Grid {
        Grid {
            width,
            height,
            points: vec![
                Color {
                    red: 0,
                    green: 0,
                    blue: 0,
                };
                (width * height) as usize
            ],
        }
    }
    fn point(&mut self, color: Color, x: u32, y: u32) {
        self.points[(x + y * self.width) as usize] = color;
    }
}

#[cfg(not(tarpaulin_include))]
fn main() {
    println!("rad");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_grid_new() {
        let grid = Grid::new(10, 10);
        assert_eq!(10, grid.width);
        assert_eq!(10, grid.height);
        assert_eq!(100, grid.points.len());
    }

    #[test]
    fn test_grid_point() {
        let mut grid = Grid::new(10, 10);
        assert_eq!(
            Color {
                red: 0,
                green: 0,
                blue: 0,
            },
            grid.points[5]
        );
        grid.point(
            Color {
                red: 255,
                green: 255,
                blue: 255,
            },
            5,
            0,
        );
        assert_eq!(
            Color {
                red: 255,
                green: 255,
                blue: 255,
            },
            grid.points[5]
        );
    }
}

File: easy/201-300/248/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"strings"
)

type Point struct {
	Red   int
	Green int
	Blue  int
}

func (p *Point) String() string {
	return fmt.Sprintf("%3d %3d %3d", p.Red, p.Green, p.Blue)
}

type Grid [][]Point

func NewGrid(columns, rows int) *Grid {
	grid := new(Grid)
	*grid = make(Grid, rows)
	for yIndex := 0; yIndex < rows; yIndex++ {
		(*grid)[yIndex] = make([]Point, columns)
		for xIndex := 0; xIndex < columns; xIndex++ {
			(*grid)[yIndex][xIndex] = Point{}
		}
	}
	return grid
}

func (g *Grid) String() string {
	var output []string
	for _, row := range *g {
		var currentRow []string
		for _, entry := range row {
			currentRow = append(currentRow, entry.String())
		}
		output = append(output, strings.Join(currentRow, "   "))
	}
	return strings.Join(output, "\n")
}

func (g *Grid) Point(color Point, x, y int) {
	(*g)[x][y] = color
}

func (g *Grid) Line(color Point, startX, startY, endX, endY int) {
	for xIndex := startX; xIndex <= endX; xIndex++ {
		(*g)[xIndex][(endY-startY)*xIndex/(endX-startX)+startY] = color
	}
}

func (g *Grid) Rect(color Point, startX, startY, width, height int) {
	for xIndex := startX; xIndex < startX+width; xIndex++ {
		for yIndex := startY; yIndex < startY+height; yIndex++ {
			fmt.Println(color, xIndex, yIndex)
			g.Point(color, xIndex, yIndex)
		}
	}
}

var zPrint = fmt.Println

func main() {
	_, _ = zPrint("hello world")
}
